### PR TITLE
first version of xml, xhtml and epub export tagging support

### PR DESCRIPTION
**Pull request by hernot**
late happy easter

Managed to add taggs and css files to support xml/xhtml export and epub generation

This is a very first version. But line numbering will still need some indeep attention. currently only tested with number location left. Which works outside floats but already beraks inside (*)
And i do expect even wors breaking for any other remaining placement (*)

(*) line numbers are not placed immediately infront or after <synaxlinegroup> trags but immediately after float tag and before end float tag.

But for some advice it should be fair and OK enough.

### CODE REVIEW

**Comment by adityam** [comment]
This is very exciting! Thanks. I should have time next week, when I'll go over both your pull requests.

**Comment by adityam** [comment]
This commit only contains the test files. Did you forget to include some files?
Regarding linenumbering: If the output is correct without linenumbering, then in the worst case linenumbering can be done at the CSS/HTML level as well. Not the ideal solution, but ....

File: tests/vim/37-SYNBOL-SYNEOL-tagging-tagging-export.tex [hunk 1]
```diff
@@ -0,0 +1,23 @@
+\setupbackend[
+export=yes,
+xhtml=yes
+]
+\setupexport[
+title={\type{SYNBOL}, \type{SYNEOL} tagging test},
+subtitle={},
+author={J.R.C.H.}
+]
+\usemodule[amsl]
```
[the remark this answers is not in the document]
Comment (hernot) [reply tests/vim/37-SYNBOL-SYNEOL-tagging-tagging-export.tex:10]: Copy paste from other source. Initially i thought to also demonstrate but that would be something for later more advanced test files if necesary at all. can go.

File: tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+This is a little test whether modifcation in \type{\SYNBOL} and \type{\SYNEOL} have any sideeffects
+when export is set to XML
+\startPython
+a=12
+def hello_world(hello)
+    print("simon says:    {}".format(hello))
+hallo_welt("how do youdo")
+\stopPython
+\stoptext
+
```
[the remark this answers is not in the document]
Comment (hernot) [reply tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex:28]: nope `numbering=yes` and the above is `numbering=no`

File: tests/vim/40-SYNBOL-SYNEOL-ln-tagging-export-float.tex [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+{%
+\startPython
+a=12
+def hello_world(hello)
+    print("simon says:    {}".format(hello))
+hallo_welt("how do youdo")
+\stopPython%
+}
+\stoptext
+
```
[the remark this answers is not in the document]
Comment (hernot) [reply tests/vim/40-SYNBOL-SYNEOL-ln-tagging-export-float.tex:37]: Nope again `numbering = yes` vs. `numbering = no` see comment above.

File: tests/vim/37-SYNBOL-SYNEOL-tagging-tagging-export.tex (hunk 1 above)
[the remark this answers is not in the document]
Comment (adityam) [reply tests/vim/37-SYNBOL-SYNEOL-tagging-tagging-export.tex:10]: Could you please remove this from all test files.

File: tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (adityam) [reply tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex:28]: Thanks. I missed that.

**Comment by adityam** [comment]
This commit only contains test files. There is no file which actually implements the tagging. Perhaps you forgot to add it to your branch. Could you please check it.

Thanks

**Comment by hernot** [comment]
Ouups missed to add them should be solved for now, squash in final commit

**Comment by adityam** [comment]
Which version of ConTeXt (`context --version`) are you using?

With MkIV 2020.01.30, the files do not compile. For example:

```
context tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex
```

gives

```
tex error       > tex error on line 8 in file tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex: ! Undefined control sequence

<recently read> \dosynchronizeexport 
                     
\vimtyping@define ...ingparameter \c!alternative }
                                                  \setvalue {\e!start raw#1}...
l.8 ...hon,tab=4,numbering=yes,numberlocation=right]
```
I get a similar error with LMTX 2020.04.27.

**Comment by hernot** [comment]
Hi 

[Inline Response to message by Aditya Mahajan, Mi, 2020-04-29 14:11
-0700: Answers and comments below corresponding cited paragraphs.]
> Which version of ConTeXt (`context --version`) are you using?
> 
> With MkIV 2020.01.30, the files do not compile. For example:
> 
> ```
> context tests/vim/38-SYNBOL-SYNEOL-ln-tagging-export.tex
> ```
> 
> gives
> 
> ```
> tex error       > tex error on line 8 in file tests/vim/38-SYNBOL-
> SYNEOL-ln-tagging-export.tex: ! Undefined control sequence
> 
> <recently read> \dosynchronizeexport 
>                      
> \vimtyping@define ...ingparameter \c!alternative }
>                                                   \setvalue {\e!start
> raw#1}...
> l.8 ...hon,tab=4,numbering=yes,numberlocation=right]
> ```
> I get a similar error with LMTX 2020.04.27.
> 
> 


I currently use 2018 tex-life als 2020 texlife from jonathonf was not
yet ready and 2019 was broken. The standalone i did not try yet as i
currently use it in production and already had a texlife setup and did
not want to switch. 

Maybe this would also solve some of my other problems. I will take care
about it in a few weeks when i hope to have finished my lecture notes
and thus safely can switch to an up to date and clean setup including
the recent release of t-vim.

Xristoph

**Comment by adityam** [comment]
> I currently use 2018 tex-life als 2020 texlife from jonathonf was not
> yet ready and 2019 was broken. The standalone i did not try yet as i
> currently use it in production and already had a texlife setup and did
> not want to switch.

Note that it is possible to install context-standalone in parallel to texlive.
In fact, you can have multiple versions of context-stanalone in parallel, so
you can keep a frozen version for long running projects but at the same time
use the latest version for the newer projects.

> Maybe this would also solve some of my other problems. I will take care
> about it in a few weeks when i hope to have finished my lecture notes
> and thus safely can switch to an up to date and clean setup including
> the recent release of t-vim.

No worries. In the meanwhile, I'll see if I can get a better understanding of
the export mechanism.

**Comment by hernot** [comment]
Ok just had some time to test home. 
The source of error is in `\setup_export` command definee by `t-syntax-groups`.
On about line 326 the `\doifmode{*export}` clause defines the method `inject_css_file` which calls at the end (following the comment) `\dosynchronizeexport`.
The last command is (still) known in 2018 release of Context but has been removed in Current release (2020).

By replacing 
`\dosynchronizeexport`
by
`\doifdenfined{dosynchronizeexport}{\dosynchronizeexport}`
The error should be solved.

But the exporter seemst to have some problems with newline characters as it complains about `possible paragraph mixup` and it does not output any tag to html/xhtml/xml strangely. Not sure if related.

**Comment by adityam** [comment]
If I change that line, I do not get any elements in the exported XML. Did you make any changes in `t-filter` or `t-vim`?

**Comment by hernot** [comment]
Thats exactly the point where I'm currently struggling with 2020 ConTeXt version. The acompanying pdf looks as should be. The only thing which i do see whith my naiv trial and error approach is by turning on export trackers is that it has some problem that the code block would sort before the first paragraph and that seems to be disliked by the exporter. 
That is why i have sent an email concerning the exporter to the mailing list. But I'm not sure if to ask development questions the normal mailing list is the best place. Especially when it comes to the need to understand how lowlevel things like the exporter work or how the document tree looks like it has to process and choke on. 

Do you have an advice which channel to use for asking these kind of questions?

**Comment by adityam** [comment]
On Mon, 18 May 2020, hernot wrote:

> That is why i have sent an email concerning the exporter to the mailing list. But I'm not sure if to ask development questions the normal mailing list is the best place. Especially when it comes to the need to understand how lowlevel things like the exporter work or how the document tree looks like it has to process and choke on.
>
> Do you have an advice which channel to use for asking these kind of questions?

The mailing list is the right channel because all the experts are on that list. But you do have to break down the problem into bite sized chunks and include easily reproducible examples.

Please don't take this as a criticism, but in my opinion your latest question did not receive any response because there was no way to reproduce the error message. You had linked to your pull request, but not stated what needs to be run in order to get the error message. Also, asking someone else to discuss internal code in a module can be difficult. If you can translate that into a smaller, stand-alone question, then there is a higher chance of positive replies.

**Comment by hernot** [comment]
Exactly, i was too clueless and disoriented to be able to get anywhere close to be able to break down to anything helpfull. I do fear for now I'm just the apprentice having to whatch and learn and not much help otherwise. Cause even with the answers you got i'm still as clue less and lost as before.

**Comment by hernot** [comment]
By the way where do i find documentatoin how to properly debug my contributions, how do is get to know what of my thoughts about context internal workings are wrong and where do i find documentation about the designprinciples of context any of its modules.

Is the mailing list the only source?

**Comment by adityam** [comment]
On Mon, 25 May 2020, hernot wrote:

>
> By the way where do i find documentatoin how to properly debug my contributions, how do is get to know what of my thoughts about context internal workings are wrong and where do i find documentation about the designprinciples of context any of its modules.

The high-level interface is documented in the manuals. There was a bug which I reported in the mailing list. Hopefully, that will be fixed in the next release.

> Is the mailing list the only source?

Yes. Mailing list + reading the source (which used to very easy to read pre-luatex, but now is more complicated because functionality is split between lua and tex).

The actual code for tagging in exports is pretty simple. I pushed a new branch called export which works with my bugfix. The only remaining issues are linenumbering and maintaining spaces (which already happens for `\starttyping`, so it is just a question of mimicking that functionality).

I am waiting for a new release with the bugfix, so that I can then use to test further.

Aditya

**Comment by adityam** [comment]
Okay, first step towards the missing features. I pushed a new version 4772457c9bb8252810653daba83fc707202a5c4a, which now fixes the spaces in export. Now looking at line numbers.

**Comment by adityam** [comment]
Linenumbering is also sorted. Was a backend issue. I am waiting for the context distribution to be updated and then we can test the export functionality.

**Comment by adityam** [comment]
See branch `export-2`. It works here for simple examples (with some local changes to strc-tag.lua and back-exp.lua). For example, the file:

```
\setupbackend[export=xml]

\usemodule[vim]

\definevimtyping[PYTHON][syntax=python]

\starttext
\startsection[title={A code listing}]
  \startparagraph
    This is a code listing
  \stopparagraph
  \startPYTHON
    # Python program listing
    def foobar
        print("Hello World")
  \stopPYTHON
  
  \startparagraph
    Now the same example with line numbering
  \stopparagraph
  \startPYTHON[numbering=yes]
    # Python program listing
    def foobar
        print("Hello World")
  \stopPYTHON
\stopsection
\stoptext
```

gets exported to

```
<?xml version="1.0" encoding="UTF-8" standalone="yes" ?>

<!--

    input filename   : test
    processing date  : 2020-05-31T14:19:54-04:00
    context version  : 2020.05.25 23:39
    exporter version : 0.35

-->

<!-- This export file is used for filtering runtime only! -->

<document context="2020.05.25 23:39" date="2020-05-31T14:19:54-04:00" file="test" language="en" version="0.35" xmlns:m="http://www.w3.org/1998/Math/MathML">
 <metadata>
 </metadata>
 <section detail="section" chain="section" implicit="1" level="3">
  <sectioncaption>
   <sectionnumber>1</sectionnumber>
   <sectiontitle>A code listing</sectiontitle>
  </sectioncaption>
  <sectioncontent>
   <paragraph>This is a code listing</paragraph>
   <vimtyping detail="PYTHON">
    <verbatimline><syntaxgroup detail="Comment"># Python program listing</syntaxgroup></verbatimline>
    <verbatimline><syntaxgroup detail="Statement">def</syntaxgroup> <syntaxgroup detail="Function">foobar</syntaxgroup></verbatimline>
    <verbatimline>    <syntaxgroup detail="Function">print</syntaxgroup>(<syntaxgroup detail="String">"</syntaxgroup><syntaxgroup detail="String">Hello World</syntaxgroup><syntaxgroup detail="String">"</syntaxgroup>)</verbatimline>
   </vimtyping>
   <paragraph>Now the same example with line numbering</paragraph>
   <vimtyping detail="PYTHON">
    <verbatimline><linenumber>1</linenumber><syntaxgroup detail="Comment"># Python program listing</syntaxgroup></verbatimline>
    <verbatimline><linenumber>2</linenumber><syntaxgroup detail="Statement">def</syntaxgroup> <syntaxgroup detail="Function">foobar</syntaxgroup></verbatimline>
    <verbatimline><linenumber>3</linenumber>    <syntaxgroup detail="Function">print</syntaxgroup>(<syntaxgroup detail="String">"</syntaxgroup><syntaxgroup detail="String">Hello World</syntaxgroup><syntaxgroup detail="String">"</syntaxgroup>)</verbatimline>
   </vimtyping>
  </sectioncontent>
 </section>
</document>
```

**Comment by hernot** [comment]
Hi adityam

Ok that i would have never sorted out, given the necessity of
contibutitions to core ConTeXt. Too many improvements there, that
making vimtyping output appear nicely in XML and E-Pub output would
just be some adoption of module code only. 

Thank You very much for taking over. 

I's there anything left, after new version of ConTeXt including your
patches has been released, for me to support you eg. helping in writing
Tests etc. ?
Otherwise if nothing is left for me to support with, I'd suggest to
close my (this) pull-request and continue with your branch.

Best
Xristoph

**Comment by adityam** [comment]
On Mon, 1 Jun 2020, hernot wrote:

> Ok that i would have never sorted out, given the necessity of
> contibutitions to core ConTeXt. Too many improvements there, that
> making vimtyping output appear nicely in XML and E-Pub output would
> just be some adoption of module code only.
>
> Thank You very much for taking over.
>
> I's there anything left, after new version of ConTeXt including your
> patches has been released, for me to support you eg. helping in writing
> Tests etc. ?

I still haven't looked at generating CSS. Since the module only provides a two built-in color schemes, I was actually thinking of including a hand coded CSS rather than an automatically generated one. But I need to look into that.

I'd also appreciate your help with testing to make sure there is no corner case.

Thanks,
Aditya

**Comment by hernot** [comment]
> I still haven't looked at generating CSS. Since the module only provides a two built-in color schemes, I was actually thinking of including a hand coded CSS rather than an automatically generated one. But I need to look into that. 

Unless changes in colour management are as radical as for tagging in latest mkiv and lmtx than in theory my suggestion on how to create css from colourscheme setup in context and hook it to css output list should still work. In other words no need for static css and for special cases users still can manually attach css through cssfile option of export. 


> I'd also appreciate your help with testing to make sure there is no corner case.

Ok please keep me posted on progress and on which dedicated tasks or issues my support would be needed and/or apriciated.

Xristoph

**Comment by adityam** [comment]
The code now works with the latest version of ConTeXt. Can you test the `export-2` branch. See `tests/41-export.tex` for a simple example. 

Please note that I have not yet added the CSS functionality (haven't had the time). I'll try to test your code and then add something appropriate. Note that ConTeXt does generate a bare-bones CSS definitions in `*-epxort/style/filename-template.css`.

**Comment by hernot** [comment]
Yes i will test with latest an report any 

Yes i know about `*-epxort/style/filename-template.css`. I decided against it for the following probably selfish reasons. 
- It is not automatically included in css used by output and has to be added manually through cssfile parameter of export by user. 
- I have no idea if it is recreated on every run or only created if not present. 
- If recreated only if missing than i consider it to be a reminder for the user to collect all his specific css stuff which is not provided by ConTeXt or any of its packages therein. 


- In case recreated every time why user still has to add manually? And what if user does not add, what does that mean for typesetting of code? 
- What side effects would occur if t-vim would automatically add to ensure for proper display of code? 

Given these  i could not fully figure i decided against utilizing this file for placing vimtyping related stuff. And yes I admit i might have been a bit lazy and lousy to not approach those via Mailing list who could have helped to obtain more clarity upon above points ;-)

**Comment by hernot** [comment]
Ok seems to work eventhough all syntaxgroups are for now maked as verbatim lines. Are there any obstacles where your expertise would be needed or is that something i could play with at least?

Just for understanding: shouldn't 41-export.tex be placed in tests/vim as it is a tests for vimtyping and not basic t-filter? or is vimtyping just the vehicle or better the environment allowing to test the underlying changes to t-filter?

**Comment by adityam** [comment]
> Ok seems to work eventhough all syntaxgroups are for now maked as verbatim lines.

What do you mean by that? Here I get:

      <vimtyping detail="PYTHON">
       <verbatimline><syntaxgroup detail="Comment"># Python program listing</syntaxgroup></verbatimline>
       <verbatimline><syntaxgroup detail="Statement">def</syntaxgroup> <syntaxgroup detail="Function">foobar</syntaxgroup></verbatimline>
       <verbatimline>    <syntaxgroup detail="Function">print</syntaxgroup>(<syntaxgroup detail="String">"</syntaxgroup><syntaxgroup detail="String">Hello World</syntaxgroup><syntaxgroup detail="String">"</syntaxgroup>)</verbatimline>
      </vimtyping>
      <paragraph>Now the same example with line numbering</paragraph>
      <vimtyping detail="PYTHON">
       <verbatimline><linenumber>1</linenumber><syntaxgroup detail="Comment"># Python program listing</syntaxgroup></verbatimline>
       <verbatimline><linenumber>2</linenumber><syntaxgroup detail="Statement">def</syntaxgroup> <syntaxgroup detail="Function">foobar</syntaxgroup></verbatimline>
       <verbatimline><linenumber>3</linenumber>    <syntaxgroup detail="Function">print</syntaxgroup>(<syntaxgroup detail="String">"</syntaxgroup><syntaxgroup detail="String">Hello World</syntaxgroup><syntaxgroup detail="String">"</syntaxgroup>)</verbatimline>
      </vimtyping>

Note that each line is marked as a `<verbatimline>` and each syntax group is marked separately inside it.

> Are there any obstacles where your expertise would be needed or is that something i could play with at least?

As far as the XML is concerned, I think that this is okay. So, please feel free to play around with this.

I haven't had the chance to look at the CSS part yet.

> Just for understanding: shouldn't 41-export.tex be placed in tests/vim as it is a tests for vimtyping and not basic t-filter? or is vimtyping just the vehicle or better the environment allowing to test the underlying changes to t-filter?

That was a mistake. Thanks for catching it. Fixed.

**Comment by hernot** [comment]
Ah yes you are right should have read carefully. And the rest is a matter of css. 
In html it appears with every elment on a single line as if firefox would only see block setting for vimtyping and not get the css div entries.



whereas in xhtml it seems to recognize the corresponding css settnings.


As far as i remember i had similiar struggles but i think it should resolve by dropping div, but I'm not quit sure how i got rid of that strange behaviour anymore. I just remember that it was gone somehow.

**Comment by adityam** [comment]
I don't have a good understanding of CSS and HTML. I'll let you figure that out :-)

⁣

**Comment by hernot** [comment]
Ok I'll refigure and pullrequest against your branch as soon as i think i do have a usable solution.

**Comment by hernot** [comment]
I'm nearly there to make things work as expected for proper display of typeset syntax in export to html, xhtml and xml. Only one little thing is to be solved where i do need your advice and guidance how to make it work without breaking anything.

The t-vim module is based upon t-filter and the outer most set of `<div>` tags for each syntax display is placed by t-filter.
It assigns to the outer most `<div>` tag on line 488 in t-filter.tex 

`\dostarttagged{\externalfilterparameter\c!taglabel}\currentexternalfilter` 

The  class name is set to the passed `taglabel` parameter  and tag detail specifier is set to the content of `\currentexternalfilter`.
For vimtyping this holds the name of the language typeset eg. "PYTHON". But for css styling the name of language has no meaning at this stage. The language has already been converted into language independent structure of `<div class="verbatimline">` and `<div class="syntaxgroup [syntaxelement]">` tags throug `2context.vim` script. 

Therefore the tag for vimtiyping should better use colorsheme name as tagdetail information like so

 `<div class="vimtyping pscolor">`

and in general 

`<div class="vimtyping [colorscheme]">`

How should i, may i or do i have to modify or extend the related line in t-filter.tex to be able to select an alternative value for the tag detail value instead of the value stored in `\currentexternalfilter`.
Thereby i want to avoid no non desired side effects on other external filters defined elsewhere occur.

**Comment by adityam** [comment]
I can think of two possibilities:

1. I add a parameter, `tagdetail` and use

     \dostarttagged{\externalfilterparameter\c!taglabel}{\externalfilterparameter\c!tagdetail}

   Then `t-vim` can set tagdetail to the name of the colorscheme (or perhaps both colorscheme and language; I have seen CSS with `class="name1 name2 name3"`).


2. We use `\startelement[taglabel][colorscheme=name-of-colorscheme, language=name-of-language]`. This is nicer because we get a more detailed XML, but it will require some refactoring as the current interface doesn't allow passing the list of key-value pairs). One also needs to globally enable `\setupelements[properties=yes]` to get the tags, and I am not sure how this interacts with other modules.

So, I am leaning towards the first, as it is trivial to implement.

**Comment by hernot** [comment]
which ever option you prefer. 

Concerning `\setupelements[properties=yes]` is that specific to t-filter and t-vim or a global for ConTeXt. In case the latter i can check if that is set automatically when `\setupbackend` and `\setupexport` commands are used.

**Comment by adityam** [comment]
I'll implement something with the first option.

`\setupelements[properties=yes]` is global for everything. The way the context backend code is written, it is not possible to localize it effect.

**Comment by hernot** [comment]
Are you sure about `\setupelements[properties=yes]`  i did not find it neither in docs nor in the standalone code and docs? Where is it mentioned?

**Comment by adityam** [comment]
Sorry,

   \setupexport[properties=yes]

or

   \setupexport[properties=prefix]

**Comment by hernot** [comment]
Ah now i get what you mean, but from reading epub-mkiv.pdf I'm not sure if that is relevant for anything beyond structures created through`\startelement \stopelement` at least that is how i would read last section 6. As it is only mentioned there and all examples shown there use `\startelement \stopelement`. And first Paragraph in this section says 

"[...]The default output reflects the structure present in the document. If that is not enough you can add your own structure [...]"

And as i view it t-filter and t-vim output is structure which is already present in the document. So I'm not sure if the properties parameter of `\setupexport` would have any effect to its tags and how it would be possible to make it have an effect. And would that be necessary at all?

**Comment by adityam** [comment]
There are two ways to implement tagging. 

The first, which we currently follow in t-filter/t-vim is

    \dostarttagging{name}{value}
    ...
    \stopstoptagging

which gets translated to

    <name detail=value>
    ...
    </name>


The other option is

    \startelement[name][option1=value1, option2=value2]
    ...
    \stopelement

By default, this gets translated to 

    <name>
    ...
    </name>

But, if \setupexport[properties=yes] is set, this gets translated to

   <name option1=value1, option2=value2>
   ...
   </name>

However, if \setupexport[properties=prefix] is set, this gets translated to

   <name prefix:option1=value1, prefix:option2=value2>
   ...
   </name>

I get the impression that \dostarttagging is for internal code while \startelement is for user code. The limitation of \dostarttagging is that we can only have one option. (There is also something which captures the inheritance structure with `chain`, but that might not be relevant for us).

If we use \startelement in t-filter, then (i) we need to enable \setupexport[properties=yes] but this might interfere with user code. For example, in a personal document, I might or might not want properties to be enabled globally. (ii) If the user sets \setupexport[properties=prefix], then the output of t-filter/t-vim gets affected. I don't like that.

I have implemented a version in `export-2` using \dostarttagged so that the output is 

    <vimtyping detail="name-of-colorscheme>
    ....
    </vimtyping>

**Comment by hernot** [comment]
So we do have the same impression that `\startelement \stopelement` is for document editors which want to add custom elements which have no corresponding item or element in other output formats like PDF. I would opt for sticking with `\starttagged \stoptagged` for now as it is more consistent with standard ConTeXt and has less side effects. Could it be that the chain attributed are filled by the third variant `\starttaggedchained` Where the detail specifier is followed by a reference to another tag? Not sure if that has any value for vimtyping. But doesn't that for now deviate a bit from the initial idea/request to get comparable results from vimtyping on default pdf output and html, xhtml and xml output.

**Comment by adityam** [comment]
> So we do have the same impression that `\startelement \stopelement` is for document editors which want to add custom elements which have no corresponding item or element in other output formats like PDF.

Yes

> I would opt for sticking with `\starttagged \stoptagged` for now as it is more consistent with standard ConTeXt and has less side effects. Could it be that the chain attributed are filled by the third variant `\starttaggedchained` Where the detail specifier is followed by a reference to another tag? Not sure if that has any value for vimtyping.

I agree.

> But doesn't that for now deviate a bit from the initial idea/request to get comparable results from vimtyping on default pdf output and html, xhtml and xml output.

Please look at the current export-2 branch and see if something is missing (apart from CSS). In my opinion the generated XML is fine. If you agree, then I can merge this to the main branch and then we can see how to include CSS.

**Comment by hernot** [comment]
Oh i have a copy of your export-2 on my repo and in my local directory already css included. Just need to run some further tests and than i would place a pull-request against your export-2 if that is ok for you. Or should i directly pullrequest my export-2 against dev?

**Comment by adityam** [comment]
Please place it against export-2. export-2 and dev have deviated, so I'll rebase locally before merging.

**Comment by hernot** [comment]
OK done by #38.
